Extend to arbitrary dimensions

rtimpe opened the pull request:
# Description
The main change is to extend the core algorithm to N dimensions.  This is mostly done with no changes to the 2d case, however in some places result matrices are transposed relative to how they were previously.  This doesn't change anything fundamental, but does produce slightly different results (i.e. in cases where an argmax is taken over the acquisition function with multiple equivalent maxima and the different iteration order causes a different maximum to be chosen).  The old behavior can be kept by setting the `match_original` flag.  In addition, the following minor changes are also included
- Remove the manifest.toml file
- Fix a bug on line 4 of plotting.jl.  I wasn't able to import the BayesianSafetyValidation model without this change - maybe the function signature changed in a more recent version of the Plots module?

# Caveats
- I'm pretty new to Julia, so please be as critical as you want.  This is particularly true for the vectorized parts of the code - there's a few places where I wrote something that feels like it might be part of the standard library.
- Plots will only work for two dimensions
- I haven't updated the code in importance_sampling.jl yet

# Testing
I've mostly tested using the dummy squares system
- Ran the 2d version, and made sure I could produce the same plots and results as the original implementation
- Ran the 3d version, and made sure it came up with a good estimate of the probability of failure.

mossr commented:
This looks fantastic, Rob. I pulled the PR and tested on my local machine—works well.

The only thing I'd suggest modifying before we pull this in is the `truth_estimate` function inside of `baselines.jl`. I'd suggest making the same edits you've done with the `input_discritization_steps` etc. to this function (this way, we can compare the output of the GP estimated failure probability with the "truth").

rtimpe commented:
Thanks for the quick review.  I haven't really looked at the `experiments` directory yet - I'll look at `baselines.jl` next

jeangud commented on src/bayesian_safety_validation.jl:
```diff
@@ -25,6 +25,9 @@ function bayesian_safety_validation(sparams, models;
                             refit_every_point=false,
+                            input_discritization_steps=200,
+                            p_estimate_discritization_steps=500,
+                            match_original=false,
```
(no action needed) I can see us deprecating this flag soon as we get used to running this new implementation.

rtimpe replied:
yeah, i think it's mainly useful for
- my testing, to make sure i haven't broken things
- if someone wants to reproduce the exact results of the paper (i.e. set the same random seed and make sure the same points are chosen by the acquisition functions, and the same GP is fit).  i'm not sure how important this use case is though

jeangud commented on src/bayesian_safety_validation.jl:
```diff
@@ -25,6 +25,9 @@ function bayesian_safety_validation(sparams, models;
                             samples_per_batch=1,
                             single_failure_mode=false,
                             refit_every_point=false,
+                            input_discritization_steps=200,
```
(nit) `discretization` typo

jeangud commented on src/bayesian_safety_validation.jl:
```diff
@@ -34,7 +37,7 @@ function bayesian_safety_validation(sparams, models;
 
             inputs = []
-            X = Matrix{Float64}(undef, 2, 0)
+            X = Matrix{Float64}(undef, length(models), 0)
```
If `len(models) == size(X)[1]` below, I would suggest initializing a `num_dimensions = length(models)` variable

jeangud commented on src/likelihood_weighting.jl:
```diff
@@ -1,21 +1,19 @@
-function p_estimate(gp, models; m=fill(500, length(models)), grid=true)
-    a₁, b₁ = models[1].range[1], models[1].range[end]
-    a₂, b₂ = models[2].range[1], models[2].range[end]
+function p_estimate(gp, models; num_steps=500, m=fill(num_steps, length(models)), grid=true)
```
Is there any case where `m` is given manually to this function? 

I see its initialization depends on two arguments so I am wondering if we could demote this from function argument to simply function variable, initializing it at the start of the function.

rtimpe replied:
it's given [here](https://github.com/sisl/BayesianSafetyValidation.jl/blob/main/src/likelihood_weighting.jl#L30).  that said, there might be other places where we're doing something similar and never passing `m` explicitly.  but i can also imagine scenarios where you would want to discretize the different input dimensions differently, even if we're not necessarily doing so currently

jeangud reviewed (approved):
A great feature to add @rtimpe, thank you!

Added minor comments if you want to have a look, nothing blocking however. I'd check with @mossr before merging.

rtimpe commented:
ok, i've updated the `truth_estimate` function, as suggested.  i also updated the `coverage` and `region_characterization` functions to allow running the `run_baselines` function.  i didn't update the baselines themselves, however, so only the `uniform` baseline will work correctly
